freqtrade/rpc/external_message_consumer.py

```diff
@@ -336,16 +366,58 @@ def _consume_analyzed_df_message(self, producer_name: str, message: WSMessageSch
 
         pair, timeframe, candle_type = key
 
+        if df.empty:
+            logger.debug(f"Received Empty Dataframe for {key}")
+            return
+
         # If set, remove the Entry and Exit signals from the Producer
         if self._emc_config.get('remove_entry_exit_signals', False):
             df = remove_entry_exit_signals(df)
 
-        # Add the dataframe to the dataprovider
-        self._dp._add_external_df(pair, df,
-                                  last_analyzed=la,
-                                  timeframe=timeframe,
-                                  candle_type=candle_type,
-                                  producer_name=producer_name)
+        logger.debug(f"Received {len(df)} candle(s) for {key}")
+
+        if len(df) >= 999:
+            # This is a full dataframe
+            # Add the dataframe to the dataprovider
+            self._dp._add_external_df(
+                pair,
+                df,
+                last_analyzed=la,
+                timeframe=timeframe,
+                candle_type=candle_type,
+                producer_name=producer_name
+            )
+
+        elif len(df) < 999:
+            # This is n single candles
+            # Have dataprovider append it to
+            # the full datafame. If it can't,
+            # request the missing candles
+            did_append, n_missing = self._dp._add_external_candle(
+                pair,
+                df,
+                last_analyzed=la,
+                timeframe=timeframe,
+                candle_type=candle_type,
+                producer_name=producer_name
+            )
+
+            if not did_append:
+                logger.debug("Holes in data or no existing df, "
+                             f"requesting {n_missing} candles "
+                             f"for {key} from `{producer_name}`")
+
+                self.send_producer_request(
+                    producer_name,
+                    WSAnalyzedDFRequest(
+                        data={
+                            "limit": n_missing,
```

Comment:
I'm ... not entirely sure this is a good idea - i'd rather go the "full DF" route - at least for now.

We assume we only hit this if there was some outage - or if the producer was too slow. Based on this - you could get the candle at :59.999 (instead of at :01). by the time we get to request a new DF (or by the time the producer get's to work on this request) - we might be missing one additional candle - so we might end up in a request/re-request loop - always falling 1 candle short - but might never actually get to a "now we have it all".

IF we take this approach - i think you'll want to request 1 overlapping candle - which will then have to be replaced (we don't know if the candle changed mid-candle ... ), which will obviously make the logic in dataprovider a bit more complex (you would now have to replace the last candle - but append subsequent candles).

Reply:
Per our conversation over Discord, we've decided to go the route of updating the `_add_external_candle` method to support overlapping dataframes. Also, always requesting `n_missing + 1`.